oepcie/yuanphil_impl/oe_ctx_impl.c

```diff
@@ -0,0 +1,221 @@
+
+	ctx -> map.num_dev = num_devices;
+	ctx -> map.devs = calloc(num_devices, sizeof(device_t));
+
+	// return to start of device_ids
+	lseek(ctx -> header.fid, 4, SEEK_SET);
+
+	for(int i = 0; i < num_devices; i++){
+		int id = read(ctx -> header.fid, &id, sizeof(int));
+		ctx -> map.devs[i] = DEVICES[id];
+	}
+
+	// calculate read offsets
+	for(int i = 1; i < num_devices; i++){
+		ctx -> map.devs[i].read_offset = ctx -> map.devs[i - 1].read_offset +
+			ctx -> map.devs[i - 1].read_size;
+	}
+
+	/*
+	 * Not implemented - Write Offsets... Aren't exactly sure what/if these are
+	 * used yet.
+	 */
+
+	 return 1;
+
+}
+
+int _oe_set_opt(oe_ctx_impl_t* ctx, int option, const char* value){
+	if(ctx -> initialized == 'T'){
+		printf("Already initialized, no more changing options\n");
+		return -1;
+	}
+
+	switch (option){
+		case OE_HEADER_STREAMPATH:
+			ctx -> header.path = realloc(ctx -> header.path, strlen(value) + 1);
+			strcpy(ctx -> header.path, value);
+			break;
+		case OE_CONFIG_STREAMPATH:
+			ctx -> config.path = realloc(ctx -> config.path, strlen(value) + 1);
+			strcpy(ctx -> config.path, value);
+			break;
+		case OE_DATA_STREAMPATH:
+			ctx -> data.path = realloc(ctx -> data.path, strlen(value) + 1);
+			strcpy(ctx -> data.path, value);
+			break;
+		default:
+			printf("Improper option passed, returning with error\n");
+			return -1;
+			break;
+	}
+
+	return 1;
+}
+
+// Allocates a buffer for value, leaves value pointing to a properly formatted
+// dynamic c-string. Returns with error if option is not defined
+int _oe_get_opt(oe_ctx_impl_t* ctx, int option, char* value){
+	switch (option){
+		case OE_HEADER_STREAMPATH:
+			value = malloc(sizeof(char) * strlen(ctx -> header.path) + 1);
+			strcpy(value, ctx -> header.path);
+			break;
+		case OE_CONFIG_STREAMPATH:
+			value = malloc(sizeof(char) * strlen(ctx -> config.path) + 1);
+			strcpy(value, ctx -> config.path);
+			break;
+		case OE_DATA_STREAMPATH:
+			value = malloc(sizeof(char) * strlen(ctx -> data.path) + 1);
+			strcpy(value, ctx -> data.path);
+			break;
+		default:
+			printf("Improper option passed, returning with error\n");
+			return -1;
+			break;
+	}
+	return 1;
+}
+
+// Checks device_id and address for register reads/writes
+int _oe_reg_prep(oe_ctx_impl_t* ctx, int device_id, int addr){
+	// Problem - What if we have multiple of the same devices?
+
+	// check for whether it is a valid device
+	if(device_id < 0 || device_id > MAX_DEVICE_ID){
+		return -1;
+	}
+
+	// For loop checks that the device is an existing device
+	int id = -1;
+	for(int i = 0; i < ctx -> map.num_dev; i++){
+		printf("Device id at index %d is %d\n", i, ctx -> map.devs[i].id); // debug
+		if(ctx -> map.devs[i].id == device_id){
+			id = ctx -> map.devs[i].id;
+		}
+	}
+
+	if(id < 0){
+		printf("(device not found)\n"); // debug
+		return -1;
+	}
+
+
+	// need some sort of error check on addr, just going to assume every device has 16 different addresses
+	if(addr > 0xF){
```

Comment:
The check will be done by the firmware by setting `oe_config_write_ack` or `oe_config_read_ack` to an error code. This is the return value of the call to `oe_write_reg()` and `oe_read_reg()`, respectively.